code-quality/validation/README.md

```diff
@@ -0,0 +1,80 @@
+<!-- Last updated 9-25-19 -->
+
+# Input Validation 
+
+Input validation is your first line of defense when creating a secure application.  
+
+## Whitelist or Blacklist? 
+
+The first concept of good input validation is whitelisting versus blacklisting. Whitelist, or inclusive validation, defines a set of valid characters while blacklist, or exclusive validation, defines a set of invalid characters to try to remove. 
+
+Input validation using blacklisting is generally more challenging to implement so it is better practice to perform validation using whitelisting. A great way of defining a whitelist for input validation is to leverage Regular Expressions(RegEx). 
+
+## RegEx Password Validation Examples (JavaScript)
+
+- Mininum length of any 8 chars but newline: 
+```javascript
+/^.{8,}$/
+```
+
+- Check for at least one uppercase character: 
+```javascript
+/([A-Z]+)/g
+```
+
+- Check for at least one lowercase character:
+```javascript
+/([a-z]+)/g
+```
+
+- Check for at least one numeric character: 
+```javascript
+/([0-9]+)/g
+```
+
+- Check for at least one special character:
+```javascript
+/([!"#$%&'()*+,\-./:;<=>?@[\]^_`{|}~])+/g
+```
+
+## RegEx Email Validation Example (JavaScript)
+
+- DISCLAIMER: RegEx email validation should not be your sole source of email validation, rather it should be used to help your UI display an `invalid email` message. Whenever possible, one should include a process to confirm a given email address (e.g. send an email for the user confirm their email address).
```

Comment:
Would appreciate further discussion on this topic, permissive vs non-permissive regex here, permissive regex makes sure we don't run into false negatives (e.g people with a correct email being unable to sign up for the systems we build) I would like to hear advocates for non-permissive regex's for sign ups defend their position on this point a little more.

Otherwise I would prefer this to say that on sign ups, we prefer permissive regex if any, and the primary source of validation and verification of correct email address to be verification emails (a standard used across the industry)

Reply:
I agree with this, but client validation provides nice UX. It doesn't prevent someone making a typo that is a well-formed email, but it does help with a typo that makes an invalid email.